csharp/ql/lib/semmle/code/csharp/security/dataflow/ZipSlipQuery.qll

```diff
@@ -3,147 +3,360 @@
+}
 
-/**
- * A sanitizer for unsafe zip extraction.
- */
-abstract class Sanitizer extends DataFlow::ExprNode { }
+abstract class UnsanitizedPathCombiner extends Expr { }
 
-/**
- * DEPRECATED: Use `Sanitizer` instead.
- *
- * A guard for unsafe zip extraction.
- */
-abstract deprecated class SanitizerGuard extends DataFlow::BarrierGuard { }
+class PathCombinerViaMethodCall extends UnsanitizedPathCombiner {
+  PathCombinerViaMethodCall() {
+    exists(MethodCall mcPathCombine |
+      mcPathCombine.getTarget().hasQualifiedName("System.IO.Path", "Combine") and
+      this = mcPathCombine
+    )
+  }
+}
 
-/** A taint tracking configuration for Zip Slip */
-class TaintTrackingConfiguration extends TaintTracking::Configuration {
-  TaintTrackingConfiguration() { this = "ZipSlipTaintTracking" }
+class PathCombinerViaStringInterpolation extends UnsanitizedPathCombiner {
+  PathCombinerViaStringInterpolation() { exists(InterpolatedStringExpr e | this = e) }
+}
 
-  override predicate isSource(DataFlow::Node source) { source instanceof Source }
+class PathCombinerViaStringConcatenation extends UnsanitizedPathCombiner {
+  PathCombinerViaStringConcatenation() { exists(AddExpr e | this = e) }
+}
 
-  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+class MethodCallGetFullPath extends MethodCall {
+  MethodCallGetFullPath() { this.getTarget().hasQualifiedName("System.IO.Path", "GetFullPath") }
+}
 
-  override predicate isSanitizer(DataFlow::Node node) { node instanceof Sanitizer }
+class GetFullPathToQualifierTaintTrackingConfiguration extends TaintTracking4::Configuration {
+  GetFullPathToQualifierTaintTrackingConfiguration() {
+    this = "GetFullPathToQualifierTaintTrackingConfiguration"
+  }
+
+  override predicate isSource(DataFlow::Node node) {
+    exists(MethodCallGetFullPath mcGetFullPath | node = DataFlow::exprNode(mcGetFullPath))
+  }
+
+  override predicate isSink(DataFlow::Node node) {
+    exists(MethodCall mc |
+      mc.getTarget() instanceof RootSanitizerMethod and
+      node.asExpr() = mc.getQualifier()
+    )
+  }
+}
+
+class PathCombinerToGetFullPathTaintTrackingConfiguration extends TaintTracking4::Configuration {
+  PathCombinerToGetFullPathTaintTrackingConfiguration() {
+    this = "PathCombinerToGetFullPathTaintTrackingConfiguration "
+  }
+
+  override predicate isSource(DataFlow::Node node) {
+    exists(UnsanitizedPathCombiner pathCombiner | node = DataFlow::exprNode(pathCombiner))
+  }
+
+  override predicate isSink(DataFlow::Node node) {
+    exists(MethodCallGetFullPath mcGetFullPath |
+      node = DataFlow::exprNode(mcGetFullPath.getArgument(0))
+    )
+  }
+}
+
+private predicate safeCombineGetFullPathSequence(MethodCallGetFullPath mcGetFullPath, Expr q) {
+  // Path.Combine THEN Path.GetFullPath is applied (with possibly arbitrary mutations)
+  exists(DataFlow::Node source, PathCombinerToGetFullPathTaintTrackingConfiguration taintTracking |
```

Comment:
I would consider specifying UnsanitizedPathCombiner instead of "exists DataFlow::node source" because that it is more specific and probably more efficient

Reply:
Nice optimization - seems to shave off a couple of seconds!